drivers/flash/soc_flash_numaker.c

```diff
@@ -0,0 +1,282 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * Copyright (c) 2023 Nuvoton Technology Corporation.
+ */
+
+#define DT_DRV_COMPAT nuvoton_numaker_fmc
+
+#include <string.h>
+#include <errno.h>
+#include <zephyr/kernel.h>
+#include <zephyr/device.h>
+#include <zephyr/drivers/flash.h>
+#include <zephyr/logging/log.h>
+#include "flash_priv.h"
+#include <NuMicro.h>
+
+LOG_MODULE_REGISTER(flash_numaker, CONFIG_FLASH_LOG_LEVEL);
+
+#define SOC_NV_FLASH_NODE DT_INST(0, soc_nv_flash)
+#define SOC_NV_FLASH_WRITE_BLOCK_SIZE DT_PROP_OR(SOC_NV_FLASH_NODE, write_block_size, 0x04)
+
+struct flash_numaker_data {
+	FMC_T *fmc;
+	struct k_sem write_lock;
+	uint32_t flash_block_base;
+};
+
+static const struct flash_parameters flash_numaker_parameters = {
+	.write_block_size = SOC_NV_FLASH_WRITE_BLOCK_SIZE,
+	.erase_value = 0xff,
+};
+
+/* Validate offset and length */
+static bool flash_numaker_is_range_valid(off_t offset, size_t len)
+{
+	/* check for min value */
+	if ((offset < 0) || (len == 0)) {
+		return false;
+	}
+
+	/* check for max value */
+	if (((int64_t) offset + len) > ((int64_t) FMC_APROM_END - FMC_APROM_BASE)) {
+		return false;
+	}
```

Comment:
Casting up works fine as long as you cast up to processor native word. Above you just ask it to engage in extra work as, probably, dropping registers to stack to make place for 64bit calculation (32 bit addition == two/three 32bit registers, 64bit addition 4/6 32bit registers), and making one more addition.
This is not the first time I can see that so I will just ask: what is wrong with something like
```
if (offset > (FMC_APROM_END - FMC_APROM_BASE) || len > (FMC_APROM_END - FMC_APROM_BASE)  || (len + offset) >  (FMC_APROM_END - FMC_APROM_BASE) ) {
}
```
?
Here is assembly for non optimized code:
```
000098f4 <decide_64>:
{
    98f4:       e92d 0fb0       stmdb   sp!, {r4, r5, r7, r8, r9, sl, fp}
    98f8:       b087            sub     sp, #28
    98fa:       af00            add     r7, sp, #0
    98fc:       6178            str     r0, [r7, #20]
    98fe:       6139            str     r1, [r7, #16]
    9900:       60fa            str     r2, [r7, #12]
        return ((int64_t)(s) + (int64_t)(o)) < total;
    9902:       697b            ldr     r3, [r7, #20]
    9904:       2200            movs    r2, #0
    9906:       603b            str     r3, [r7, #0]
    9908:       607a            str     r2, [r7, #4]
    990a:       693b            ldr     r3, [r7, #16]
    990c:       17da            asrs    r2, r3, #31
    990e:       469a            mov     sl, r3
    9910:       4693            mov     fp, r2
    9912:       e9d7 2300       ldrd    r2, r3, [r7]
    9916:       4611            mov     r1, r2
    9918:       eb11 040a       adds.w  r4, r1, sl
    991c:       eb43 050b       adc.w   r5, r3, fp
    9920:       68fb            ldr     r3, [r7, #12]
    9922:       2200            movs    r2, #0
    9924:       4698            mov     r8, r3
    9926:       4691            mov     r9, r2
    9928:       4544            cmp     r4, r8
    992a:       eb75 0309       sbcs.w  r3, r5, r9
    992e:       bfb4            ite     lt
    9930:       2301            movlt   r3, #1
    9932:       2300            movge   r3, #0
    9934:       b2db            uxtb    r3, r3
}
    9936:       4618            mov     r0, r3
    9938:       371c            adds    r7, #28
    993a:       46bd            mov     sp, r7
    993c:       e8bd 0fb0       ldmia.w sp!, {r4, r5, r7, r8, r9, sl, fp}
    9940:       4770
```
vs
```
00009942 <decide_if>:
{
    9942:       b480            push    {r7}
    9944:       b085            sub     sp, #20
    9946:       af00            add     r7, sp, #0
    9948:       60f8            str     r0, [r7, #12]
    994a:       60b9            str     r1, [r7, #8]
    994c:       607a            str     r2, [r7, #4]
        return (o < total && s < total && (s + o) < total);
    994e:       68bb            ldr     r3, [r7, #8]
    9950:       687a            ldr     r2, [r7, #4]
    9952:       429a            cmp     r2, r3
    9954:       d90b            bls.n   996e <decide_if+0x2c>
    9956:       68fa            ldr     r2, [r7, #12]
    9958:       687b            ldr     r3, [r7, #4]
    995a:       429a            cmp     r2, r3
    995c:       d207            bcs.n   996e <decide_if+0x2c>
    995e:       68ba            ldr     r2, [r7, #8]
    9960:       68fb            ldr     r3, [r7, #12]
    9962:       4413            add     r3, r2
    9964:       687a            ldr     r2, [r7, #4]
    9966:       429a            cmp     r2, r3
    9968:       d901            bls.n   996e <decide_if+0x2c>
    996a:       2301            movs    r3, #1
    996c:       e000            b.n     9970 <decide_if+0x2e>
    996e:       2300            movs    r3, #0
    9970:       f003 0301       and.w   r3, r3, #1
    9974:       b2db            uxtb    r3, r3
}
    9976:       4618            mov     r0, r3
    9978:       3714            adds    r7, #20
    997a:       46bd            mov     sp, r7
    997c:       bc80            pop     {r7}
    997e:       4770            bx      lr
```
Just saving/restoring registers on entry/exit takes extra twelve word transfers.

Reply:
My viewpoint is to keep the result of "offset+len" and not missing any bit caused by overflow.
In fact, your suggestion is working for this case and got better code optimization.

Reply:
I applied your suggestion for better performance and is enough for this validation.
Update at commit 5baf96a32bb7.
Thanks.